fdbserver/worker.actor.cpp

```diff
@@ -3363,7 +3363,14 @@ ACTOR Future<Void> monitorLeaderWithDelayedCandidacy(
 	state Future<Void> monitor = monitorLeaderWithDelayedCandidacyImpl(connRecord, currentCC);
 	state Future<Void> timeout;
 
-	wait(recoveredDiskFiles);
+	// When encryption is enabled, and if the worker is a storage node, it needs to get the cluster controller
+	// interface in order to further obtain the updated EncryptKeyProxy interface to open local storage engine.
+	// Waiting for recoveredDiskFiles here will prevent the worker receive cluster controller interface,
+	// and create a deadlock.
+	// TODO(yiwu): can we remove the wait even when encryption is not enabled?
+	if (!SERVER_KNOBS->ENABLE_ENCRYPTION) {
```

Comment:
How about when we move away from KNOBS and encryption feature is controlled by `db-config`?

Reply:
I don't know enough about when db config is available to the workers on cluster recovery. But one option would be we just remove the wait regardless of encryption state.

Reply:
My concern is as/when `db-config` switch is done, the code needs to read the `db-config` instead of KNOBs and not sure if complicates the matter more then :)